### DISCUSSION TITLE
[Proposal] Add generic constraint 'notabstract'

### DISCUSSION THREAD

**Discussion by iikuzmychov** [dotnet/csharplang#7311] [Language Ideas]
### Background and motivation

Sometimes we need to pass a type of class that will be instantiated using reflection. We can't instantiate an abstract class, so it will be useful to have the possibility to write something like this:
```cs
where T : notabstract
```

All that this constraint should do is disallow a developer to pass the abstract class as `T`.

Of course, we have the `new()` constraint, but it's not very useful in similar situations, because we can't use it when we have some constructor arguments or required members.

I've selected `notabstract` name instead of `concrete` or something else because it will be similar to the existing `notnull` constraint name.

### Implementation details

We need to prevent the possibility to use `notabstract` with `struct` or constraint because it has no sense.

### Why it's important?

One of the reasons, why many developers love C# is that it has strict semantic. My proposal isn't a high-priority task but any possibility of making C# more strict has sense IMO.

**Comment by 333fred** [comment]
I do want us to look more deeply at first class support for factories and generic construction in the language. Until then, though, we have a pretty good solution for this in static abstracts that requires no reflection:

```cs
public interface IConstructible<T> where T : IConstructible<T>
{
    static abstract T Construct(/* Whatever args you need to be able to pass */);
}
```

**Comment by huoyaoyuan** [comment]
Features that **restricts** usage of symbols and doesn't provide additional features under the restriction can be implemented solely by analyzer. Just report an error when inappropriate usage detected.

**Comment by Joe4evr** [comment]
[Shameless plug.](https://www.nuget.org/packages/JiiLib.Constraints/)
